### ISSUE TITLE
Descriptive Stats for Segments

### ISSUE THREAD

**Issue by mschrothcp** [hglanz/AcousticsApp-Summer2019#12]
@hglanz @michalg04 

Here are the descriptive stats that could be useful for each of the segments:
Max frequency, min frequency, (and thus frequency range: max f - min f),
Max time, min time, and thus time range
Peak frequency: frequency with the largest amplitude

Of course knowing what type of call would help, but I don't think that is achievable at this point.

**Comment by michalg04** [comment]
The new segmentation method I'm using is basing the segmentation on amplitude changes (the screenshots below is what I have so far). I've been working on finding the frequency range using the spectrogram from our exploration tab, since this function doesn't give any frequency details. I'm running into some problems with the frequency and having to use the -29 trick as well as the minimum frequency always showing as 0. Do you have any suggestions? Does what I have so far look good? @mschrothglanz @hglanz 
<img width="956" alt="Screen Shot 2019-07-25 at 1 17 16 PM" src="https://user-images.githubusercontent.com/35315239/61905764-b722e180-aede-11e9-9c30-f92896422e53.png">
<img width="933" alt="Screen Shot 2019-07-25 at 1 17 28 PM" src="https://user-images.githubusercontent.com/35315239/61905771-b9853b80-aede-11e9-83f6-568f8ee6d367.png">

**Comment by hglanz** [comment]
@michalg04 this looks amazing! I'm not sure that we care about the peaks/bursts as much as the segments, so for right now let's take the peaks/bursts table off of this page.

Other notes:
- So you have min time, max time, time range. Let's add units to the headers of these columns. 
- We want min freq, max freq, and freq range as well. @mschrothglanz has suggested another way to get at these: run the segmentation method on the amplitude (y) vs. frequency (x) content. In other words, do what you've already done above but with frequency on the x-axis instead of time. Then take the frequency information from the segments from this analysis. Is this possible based on the data object you get from reading in the WAV file?
- Once you have the time and frequency boundaries for each segment, the peak frequency for each segment should hopefully be straightforward...the frequency in the time/frequency box for that segment with the highest amplitude.

Let us know if you have questions or how this goes! Sorry for the delay in response.

**Comment by michalg04** [comment]
@hglanz I'm not sure if I can give the segmentation method that specification change but I'll keep working on it and see what I can get for the amplitude (y) vs. frequency (x). Will make the other changes as well. Thanks!

**Comment by michalg04** [comment]
I used a similar idea to what @mschrothglanz suggested. Using the spec() function I got a graph that was amplitude (y) vs. frequency (x)  and just extracted min and max frequency for a given time frame (the time frame was determined by the segment). I also found a function that returns dominant frequency - dfreq() i.e. the frequency of highest amplitude, which sounded like what we were looking for. This all works much faster than the previous function and barely takes any time to load. Let me know what you think! @hglanz @mschrothglanz 
<img width="627" alt="Screen Shot 2019-07-29 at 5 59 51 PM" src="https://user-images.githubusercontent.com/35315239/62092654-d1382900-b22a-11e9-80b7-9e987341f69b.png">

**Comment by mschrothcp** [comment]
@michalg04 This seems great! I'm a little confused that the min frequencies printed are all 0..... Would it be possible to show the "box" around the segment on the spectrogram graph? That way, it can be easily seen that the min frequency is 0. What do you think @hglanz?

**Comment by hglanz** [comment]
So it sounds like there are (at least) two things to address in @mschrothglanz comment that I agree with:

1. It sounds like it would be nice to also see the segmentation results on the spectrogram. This could mean that it would be nice to reproduce the spectrogram on the segmentation tab. With this, it could be pretty cool to be able to click on a segment (row in the table) and have the box appear on the spectrogram. Reproducing the spectrogram feels a little redundant right now, but I still think we should try this out and then revise after. 

2. Perhaps only after (1) is done (?), resolve the min freq being 0 “issue”.

**Comment by mschrothcp** [comment]
That seems like a good game plan to me.

**Comment by michalg04** [comment]
So since I'm getting min/max frequency from a graph that is amplitude (y) vs. frequency (x), it's using a spectrum function and not a spectrogram. Should that graph be plotted so we can see min and max points clearly like where the min freq is 0 or do you want just the spectrogram from the exploration page? I think it would be more clear if we use the spectrum as to why min/max are what they are. @hglanz @mschrothglanz

**Comment by mschrothcp** [comment]
@michalg04 Hmmm, that's interesting. I would think that the spectrogram would be the more useful tool for graphing each segment in a box. I understand your point about the spectrum function, but it seems most useful if the viewer can see everything on one graph and the spectrogram graph allows for this. I'm thinking, if you can, to start with the spectrogram and then include the spectrum if need be. Does that work? @hglanz

**Comment by hglanz** [comment]
Yeah, sounds good to start with.

**Comment by michalg04** [comment]
Just to clarify, I should still be finding min/max frequency from the spectrum graph that is amplitude (y) vs. frequency (x), but plotting a spectrogram and using the time axis to find each segment on the spectrogram? Also, for the box, I assume that the length would be the time length of the segment but would the height be frequency range? @mschrothglanz @hglanz

**Comment by mschrothcp** [comment]
If I'm understanding you correctly, that all sounds good to me!

**Comment by michalg04** [comment]
@hglanz @mschrothglanz 
Here is an example with the dolphin file and whale files (the highlighted rows were clicked on):
<img width="657" alt="Screen Shot 2019-07-31 at 11 04 45 AM" src="https://user-images.githubusercontent.com/35315239/62318629-1ccc1c00-b451-11e9-8352-fe765ca2731a.png">
The dolphin min freq aren't exactly 0 like the whale's so we can see the min freq isn't always just 0.
<img width="627" alt="Screen Shot 2019-07-31 at 11 04 00 AM" src="https://user-images.githubusercontent.com/35315239/62318637-1f2e7600-b451-11e9-9034-a4302924ccbf.png">
These are actually boxes and not just two lines as we can see if we don't set the limits like we do in the exploration page. 
<img width="757" alt="Screen Shot 2019-07-31 at 11 37 52 AM" src="https://user-images.githubusercontent.com/35315239/62318709-408f6200-b451-11e9-9311-1a8cf058b2d8.png">
The freq min and max in the exploration page is calculated using spectro() function (spectrogram) and then the min is set to 0 and the max is set with the -29 trick, whereas I get the min and max frequency from the spec() function (spectrum) which gives a wider range of max frequency which is why the red boxes appear as red lines. If we wanted to boxes to be a bit tighter maybe we can decide on an amplitude threshold? or I could go with -29 again? 
But overall the segmentation tab is working nicely and it's pretty fast so far and no weird bugs!

**Comment by hglanz** [comment]
@michalg04 this looks so awesome!!! Ideas:

- for our purposes, could you put some histograms together of the amplitude within each (or just some) of those segments? That way we might be able to figure out a threshold in an informed way. 

- @mschrothglanz says it could be neat to display all of the segment boxes as soon as the method is run, but highlight them somehow when clicked on....what do you think?

**Comment by michalg04** [comment]
@hglanz Most of the segments for dolphin and whale respectively looked like this: 
<img width="710" alt="Screen Shot 2019-08-01 at 2 49 04 PM" src="https://user-images.githubusercontent.com/35315239/62331690-82c89b80-b471-11e9-8722-a5be9ccefc8c.png">
<img width="702" alt="Screen Shot 2019-08-01 at 2 48 53 PM" src="https://user-images.githubusercontent.com/35315239/62331692-852af580-b471-11e9-9c3e-8511ee114d3d.png">
I'm a little bit confused about the color scale since we can clearly see here and in the spectrum that the amplitude can range beyond -30.
I played around with it a bit and got to a threshold (-35) that works well and encloses the segments more but I still can't really explain why -35 works if the scale shows that it only goes to -30. Here is what it looks like with the threshold:
<img width="661" alt="Screen Shot 2019-08-01 at 2 50 13 PM" src="https://user-images.githubusercontent.com/35315239/62331791-e0f57e80-b471-11e9-9a43-e7edcf0fac4b.png">
<img width="626" alt="Screen Shot 2019-08-01 at 2 51 08 PM" src="https://user-images.githubusercontent.com/35315239/62331797-e5ba3280-b471-11e9-8f30-beac5ed599d1.png">

Lastly, I think what @mschrothglanz suggested looks good:
<img width="751" alt="Screen Shot 2019-08-01 at 3 29 07 PM" src="https://user-images.githubusercontent.com/35315239/62332279-ba384780-b473-11e9-8d33-8a7da3a885e3.png">

**Comment by hglanz** [comment]
@michalg04 in the whale histogram are you able to see the cluster of really negative amplitudes (-120 to -90) in the spectrogram as some color? That far left hump is puzzling me a lot. 

Otherwise everything sounds and looks great!

**Comment by michalg04** [comment]
@hglanz I made boxes around -90 to -120 amplitudes and they just look white. We can't see a lot of them since the limits we put on the spectrogram don't go pass frequency of 3 for the whale.
<img width="724" alt="Screen Shot 2019-08-02 at 6 59 00 PM" src="https://user-images.githubusercontent.com/35315239/62405858-cbf31b00-b557-11e9-89af-9808a68ddf71.png">

**Comment by hglanz** [comment]
@michalg04 ok I think this is a good sign that your threshold of -35 or around there is probably good. I think you’re going down good paths :)
